[gRPC] Migrate raylet client implementation to grpc

jiangzihao2009 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## What do these changes do?
At present, raylet use boost asio library to listen message from raylet client. This pr substitute asio for grpc.
1. Change raylet client to grpc client. Using asynchronous call for most requests and the requests with a reply will be sent by a synchronous grpc call.
2. Change raylet server to grpc server and handle all the request in node manager service.


## Related issue number

<!-- For example: "Closes #1234" -->

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15110/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15116/
Test FAILed.

pcmoritz commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -218,8 +220,8 @@ void SchedulingQueue::RemoveTasksFromQueue(
     if (queue->RemoveTask(task_id, removed_tasks)) {
-      RAY_LOG(DEBUG) << "Removed task " << task_id << " from "
-                     << GetTaskStateString(task_state) << " queue";
+      RAY_LOG(INFO) << "Removed task " << task_id << " from "
```
These should be switched back to DEBUG

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15188/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15275/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1586/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1607/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15298/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15305/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15310/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15311/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1614/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1619/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1620/
Test FAILed.

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -337,6 +335,18 @@ void NodeManager::Heartbeat() {
+    for (const auto &worker : dead_workers) {
+      RAY_LOG(INFO) << "Worker " << worker->GetWorkerId()
+                    << " dead because of timeout, pid: " << worker->Pid();
+      ProcessDisconnectClientMessage(worker->GetWorkerId(), worker->IsKilling());
```
would `IsKilled` be better here?

jiangzihao2009 replied:
Use killing because we just sent the sigterm message, we only can confirm it when raylet server lost heartbeats.

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -931,65 +853,75 @@ void NodeManager::ProcessGetTaskMessage(
   cluster_resource_map_[local_client_id].SetLoadResources(
       local_queues_.GetResourceLoad());
+
+  // Reply would be sent in when already assigned task to a worker.
```
could you please rephrase this?

zhijunfu replied:
e.g. Reply would be sent after raylet assigns a task to the worker.

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -931,65 +853,75 @@ void NodeManager::ProcessGetTaskMessage(
+    const rpc::DisconnectClientRequest &request, rpc::DisconnectClientReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  const WorkerID worker_id = WorkerID::FromBinary(request.worker_id());
+  bool intentional_disconnect = request.intentional();
```
not sure if I get this correctly, but if this is a `disconnect` request from a worker, then it should be `intentional`?

jiangzihao2009 replied:
yes, it's not used here and i add it just because the previous implementation.

raulchen replied:
let's just remove it.

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -931,65 +853,75 @@ void NodeManager::ProcessGetTaskMessage(
-    if (worker->IsDead()) {
-      // If the worker was killed by us because the driver exited,
-      // treat it as intentionally disconnected.
-      intentional_disconnect = true;
```
why this is removed?

jiangzihao2009 replied:
I think it's not used, because we handle the process disconnect in the heartbeat function and set it as arguments outside this func.

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1050,36 +983,42 @@ void NodeManager::ProcessDisconnectClientMessage(
+void NodeManager::HandleSubmitTaskRequest(const rpc::SubmitTaskRequest &request,
+                                          rpc::SubmitTaskReply *reply,
+                                          rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(DEBUG) << "Handle submit task request.";
```
at `task_id` into the log? otherwise we're not able to distinguish which task it is.

jiangzihao2009 replied:
It seems there is a very detailed log is in SubmitTask function

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1120,63 +1063,62 @@ void NodeManager::ProcessWaitRequestMessage(
+void NodeManager::HandlePushErrorRequest(const rpc::PushErrorRequest &request,
+                                         rpc::PushErrorReply *reply,
+                                         rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(DEBUG) << "Handle push error request.";
```
add some more info to this log?

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1287,13 +1226,83 @@ void NodeManager::ProcessSetResourceRequest(
+}
+
+/// Handle a `FreeObjectsInObjectStore` request.
+void NodeManager::HandleFreeObjectsInObjectStoreRequest(
```
This is a bit too longer, maybe just rename to `HandleFreeObjectsRequest`?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1642/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15336/
Test FAILed.

raulchen commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -6,14 +6,20 @@ namespace ray {
   // Register all the services to this server.
   if (services_.size() == 0) {
-    RAY_LOG(WARNING) << "No service is found when start grpc server " << name_;
+    RAY_LOG(WARNING) << "No service is found when start " << name_ << " grpc server.";
```
```suggestion
    RAY_LOG(WARNING) << "No service is found when starting " << name_ << " grpc server.";
```

raulchen replied:
should this be fatal?

raulchen commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -23,7 +29,9 @@ void GrpcServer::Run() {
-  RAY_LOG(INFO) << name_ << " server started, listening on port " << port_ << ".";
+  RAY_LOG(INFO) << name_
+                << " server started, input listening server address: " << server_address
+                << ", actually listening port: " << port_;
```
just `name_  << " server started, listening on " << actual_server_address`?

raulchen commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -76,6 +84,7 @@ void GrpcServer::PollEventsFromCompletionQueue() {
       // Second, server has sent reply to client and failed, the server call's status is
       // SENDING_REPLY
       if (server_call->GetState() == ServerCallState::SENDING_REPLY) {
+        RAY_LOG(INFO) << "Received reply failed call.";
```
Is this still needed?

raulchen commented on src/ray/rpc/grpc_server.h:
```diff
@@ -35,6 +35,11 @@ class GrpcServer {
   GrpcServer(const std::string &name, const uint32_t port)
       : name_(name), port_(port), is_closed_(true) {}
 
+  GrpcServer(const std::string &name, const std::string &unix_socket_path)
```
Document this constructor, and also update the other constructor's documentation.
E.g., `Construct a gRPC server that listens on a TCP port`.

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -912,96 +820,112 @@ void NodeManager::HandleDisconnectedActor(const ActorID &actor_id, bool was_loca
+  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(worker_id);
+  RAY_LOG(DEBUG) << "Handle get task request, worker id " << worker_id << " pid "
+                 << worker->Pid();
+  if (worker->IsKilling()) return;
```
should call `send_reply_callback`?

jiangzihao2009 replied:
good catch

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -912,96 +820,112 @@ void NodeManager::HandleDisconnectedActor(const ActorID &actor_id, bool was_loca
+  send_reply_callback(Status::OK(), nullptr, nullptr);
+}
+
+void NodeManager::ProcessDisconnectClientMessage(const WorkerID &worker_id,
```
rename this to `HandleClientDisconnected`, because there's no `DisconnectClientMessage` now.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15334/
Test PASSed.

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1279,15 +1203,22 @@ void NodeManager::ProcessSetResourceRequest(
+  gcs_client_->client_table().GetClient(client_id, data);
+  // Replace the resource vectors with the resource deltas from the message.
+  // RES_CREATEUPDATE and RES_DELETE entries in the ClientTable track changes
+  // (deltas) in the resources
```
this comment is out-dated.

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1279,15 +1203,22 @@ void NodeManager::ProcessSetResourceRequest(
+  // (deltas) in the resources
+  data.add_resources_total_label(resource_name);
+  data.add_resources_total_capacity(capacity);
+  // Set the correct flag for entry_type
```
also here.

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1299,13 +1230,83 @@ void NodeManager::ProcessSetResourceRequest(
+  auto data_shared_ptr = std::make_shared<ClientTableData>(data);
+  auto client_table = gcs_client_->client_table();
+  RAY_CHECK_OK(gcs_client_->client_table().Append(
+      JobID::Nil(), client_table.client_log_key_, data_shared_ptr, nullptr));
```
This is out-dated as well

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1774,40 +1798,53 @@ bool NodeManager::AssignTask(const Task &task) {
 void NodeManager::FinishAssignedTask(Worker &worker) {
   TaskID task_id = worker.GetAssignedTaskId();
-  RAY_LOG(DEBUG) << "Finished task " << task_id;
+  RAY_LOG(DEBUG) << "Finished task " << task_id << " on pid " << worker.Pid();
```
also print worker id

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -2056,7 +2093,8 @@ void NodeManager::HandleTaskReconstruction(const TaskID &task_id) {
         RAY_CHECK(lineage_cache_.ContainsTask(task_id))
-            << "Task metadata not found in either GCS or lineage cache. It may have been "
+            << "Task metadata not found in either GCS or lineage cache. It may "
+               "have been "
```
this looks weird. merge these two short lines

raulchen commented on src/ray/raylet/worker.h:
```diff
@@ -20,19 +23,19 @@ namespace raylet {
-  void MarkDead();
-  bool IsDead() const;
+  void MarkAsKilling();
+  bool IsKilling() const;
```
`MarkAsBeingKilled` and `IsBeingKilled`, and documentation

raulchen commented on src/ray/raylet/worker.h:
```diff
@@ -56,28 +57,33 @@ class Worker {
   ResourceIdSet ReleaseTaskCpuResources();
   void AcquireTaskCpuResources(const ResourceIdSet &cpu_resources);
 
+  int HeartbeatTimeout() { return ++heartbeat_timeout_times_; }
```
`TickHeartbeatTimer`?

raulchen commented on src/ray/raylet/worker.h:
```diff
@@ -56,28 +57,33 @@ class Worker {
+  /// Reply of get task request is not sent in function HandleGetTaskRequest. Should block
+  /// the worker and handle reply in AssignTask function.
+  /// Set the reply and callback when the worker sends get task request to raylet.
+  void SetGettingTaskRequest(rpc::GetTaskReply *reply,
```
`SetGetTaskReplyAndCallback`

raulchen commented on src/ray/raylet/worker.h:
```diff
@@ -88,11 +94,20 @@ class Worker {
   ResourceIdSet task_resource_ids_;
   std::unordered_set<TaskID> blocked_task_ids_;
+  /// Record worker heartbeat timeout times.
+  int heartbeat_timeout_times_;
```
```
/// How many heartbeats have been missed for this worker.
int num_missed_heartbeats_;
```

raulchen commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -380,21 +382,44 @@ std::string WorkerPool::DebugString() const {
   }
 }
 
+void WorkerPool::GetDeadWorkers(int timeout_times_limit,
```
```suggestion
void WorkerPool::TickHeartbeatTimer(int max_missed_heartbeats,
```

jiangzihao2009 commented on src/ray/raylet/node_manager.cc:
```diff
@@ -741,147 +746,50 @@ void NodeManager::DispatchTasks(
+  // Move the ASSIGNED task to the RUNNING queue so that we remember that we have
+  // it queued locally. Should move task outside assign task function because remove
+  // task might influence the iterator.
+  local_queues_.MoveTasks(assigned_task_ids, TaskState::READY, TaskState::RUNNING);
```
Move task combine remove and queue task. The remove operation may break the iterator

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15340/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15341/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1644/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15342/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1648/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1649/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1650/
Test FAILed.

stephanie-wang commented on src/ray/raylet/worker.cc:
```diff
@@ -110,10 +107,15 @@ void Worker::AcquireTaskCpuResources(const ResourceIdSet &cpu_resources) {
 
+void Worker::SetGetTaskReplyAndCallback(rpc::GetTaskReply *reply,
+                                        rpc::SendReplyCallback send_reply_callback) {
+  reply_ = reply;
```
Can we add a check here that both of these are `nullptr`? Also, is it true that this should only be called if `UsePush()`?

jiangzihao2009 replied:
Yes, it's sure that they are both nullptr.
And this function is only called when UsePush() returns false. We should not call send_reply_callback here because we want to block the raylet client until the task has been assigned to the worker. So we save send_reply_callback here and call it in AssignTask function.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15344/
Test FAILed.

stephanie-wang commented on src/ray/raylet/worker.cc:
```diff
@@ -110,10 +107,15 @@ void Worker::AcquireTaskCpuResources(const ResourceIdSet &cpu_resources) {
 }
 
+void Worker::SetGetTaskReplyAndCallback(rpc::GetTaskReply *reply,
+                                        rpc::SendReplyCallback send_reply_callback) {
```
```suggestion
                                        const rpc::SendReplyCallback &&send_reply_callback) {
```

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -741,147 +746,50 @@ void NodeManager::DispatchTasks(
-      return;
-    }
-  }
+  // Move the ASSIGNED task to the RUNNING queue so that we remember that we have
```
```suggestion
  // Move the ASSIGNED task to the RUNNING queue.
```

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -912,96 +820,114 @@ void NodeManager::HandleDisconnectedActor(const ActorID &actor_id, bool was_loca
+    send_reply_callback(Status::Invalid("WorkerBeingKilled"), nullptr, nullptr);
+    return;
   }
+  RAY_CHECK(worker && !worker->UsePush());
```
I would move the check for `worker` out to an above line, since it would already trigger a segfault at the `worker->IsBeingKilled()` line above.

jiangzihao2009 replied:
Good catch, thanks. I add worker->IsBeingKilled() sentences after this RAY_CHECK and forgot it.

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1202,44 +1139,35 @@ void NodeManager::ProcessPrepareActorCheckpointRequest(
-              }
-            });
+        reply->set_checkpoint_id(checkpoint_id.Binary());
+        send_reply_callback(Status::OK(), nullptr, []() {
```
```suggestion
        send_reply_callback(Status::OK(), nullptr, /*failure_callback=*/[]() {
```

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1132,63 +1068,64 @@ void NodeManager::ProcessWaitRequestMessage(
+              }
+            },
+            /* failure callback */
+            [this, worker_id]() { ProcessDisconnectClientMessage(worker_id); });
```
It seems like this isn't necessary anymore since all worker failures will be handled with heartbeats. Let's just use `nullptr` instead to be consistent with the other worker failures. Another option is we *always* call `ProcessDisconnectClientMessage` if a reply fails.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -189,7 +163,20 @@ class RayletClient {
 
  private:
-  const ClientID client_id_;
+  /// Try to register client in raylet, we would retry serveral time to
```
```suggestion
  /// Try to register client in raylet, we will retry several times to
```

stephanie-wang commented on src/ray/rpc/raylet/raylet_server.h:
```diff
@@ -0,0 +1,258 @@
+/// `src/ray/protobuf/raylet.proto`.
+class RayletServiceHandler {
+ public:
+  virtual void HandleRegisterClientRequest(const RegisterClientRequest &request,
```
A lot of these handlers have to call `send_reply_callback` with the same arguments at the end of the function, which feels brittle. Do you think we can have a wrapper function to call the `send_reply_callback` for handlers that fit this pattern? I'm okay with doing this in a future PR.

jiangzihao2009 replied:
You mean a wrapper function like `ReplyOk(const SendReplyCallback &send_reply_callback) { send_reply_callback(Stauts::OK(), nullptr, nullptr); }`?

stephanie-wang replied:
Yes, something like that, just so we don't have to add `send_reply_callback` to the signature of functions like `HandleRegisterClientRequest` every time.

raulchen replied:
I think `send_reply_callback` needs to be added to the signature. because it's a lambda passed from `HandleRequestImpl`. It captures some local information about this request.

stephanie-wang replied:
Right, I'm suggesting that we remove `send_reply_callback` from the signature of `HandleRegisterClientRequest` and wrap `HandleRegisterClientRequest` in a lambda that calls `send_reply_clalback`. So something like this:
```
auto wrapped_handler = [](const rpc::RegisterClientRequest &request,
                                           rpc::RegisterClientReply *reply,
                                           rpc::SendReplyCallback send_reply_callback) {
  HandleRegisterClientRequest(request);
  send_reply_callback(Status::OK(), nullptr, nullptr);
};
```
We can probably create a templated helper function that returns the wrapped handler.

stephanie-wang commented on src/ray/rpc/raylet/raylet_server.h:
```diff
@@ -0,0 +1,258 @@
+      std::vector<std::pair<std::unique_ptr<ServerCallFactory>, int>>
+          *server_call_factories_and_concurrencies) override {
+    // Initialize the factory for `RegisterClient` requests.
+    std::unique_ptr<ServerCallFactory> register_client_call_factory(
```
It would be great to have a macro for these patterns, but if it's too much it can wait for a future PR.

jiangzihao2009 replied:
Good advise, I think I should open a PR for it.

stephanie-wang commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -380,21 +382,44 @@ std::string WorkerPool::DebugString() const {
 }
 
+void WorkerPool::TickHeartbeatTimer(int max_missed_heartbeats,
+                                    std::vector<std::shared_ptr<Worker>> *dead_worker) {
```
```suggestion
                                    std::vector<std::shared_ptr<Worker>> *dead_workers) {
```

stephanie-wang commented on src/ray/raylet/worker.h:
```diff
@@ -88,11 +95,20 @@ class Worker {
   rpc::ClientCallManager &client_call_manager_;
   /// The rpc client to send tasks to this worker.
   std::unique_ptr<rpc::WorkerTaskClient> rpc_client_;
+  /// Reply of the `GetTask` request.
```
Can you write in this comment why we have to store the reply and callback for `GetTask` requests?

jiangzihao2009 replied:
It seems Hao Chen has added a detailed comment in SetGetTaskReplyAndCallback definition to explain why we have to store them.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,385 @@
+}
+
+void RayletClient::TryRegisterClient(int times) {
+  // We should block here until register succeed.
```
```suggestion
  // We should block here until register succeeds.
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15347/
Test FAILed.

edoakes commented on src/ray/raylet/worker_pool.h:
```diff
@@ -129,6 +128,14 @@ class WorkerPool {
+  ///
+  /// \param[in] max_missed_heartbeats The maximum number of heartbeats that can be
+  /// missed before a worker times out.
+  /// \param[out] dead_workers Workers that have been dead.
```
Does this mean workers that were already dead, all workers that have ever died or those that have died since the last time this was called? Please make the comment more specific.

edoakes commented on src/ray/raylet/worker_pool_test.cc:
```diff
@@ -78,19 +78,9 @@ class WorkerPoolTest : public ::testing::Test {
+    WorkerID worker_id = WorkerID::FromRandom();
     return std::shared_ptr<Worker>(
-        new Worker(pid, language, -1, client, client_call_manager_));
+        new Worker(worker_id, pid, -1, language, client_call_manager_));
```
Should have already been there, but inline comment for what the -1 means please.

edoakes commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -6,14 +6,20 @@ namespace ray {
 void GrpcServer::Run() {
-  std::string server_address("0.0.0.0:" + std::to_string(port_));
+  std::string server_address;
+  // Set unix domain socket or tcp address
```
```suggestion
  // Set unix domain socket or tcp address.
```

edoakes commented on src/ray/rpc/grpc_server.cc:
```diff
@@ -6,14 +6,20 @@ namespace ray {
   // Register all the services to this server.
   if (services_.size() == 0) {
-    RAY_LOG(WARNING) << "No service is found when start grpc server " << name_;
+    RAY_LOG(FATAL) << "No service is found when starting " << name_ << " grpc server.";
```
```suggestion
    RAY_LOG(FATAL) << "No service found when starting " << name_ << " grpc server.";
```

edoakes commented on src/ray/rpc/grpc_server.h:
```diff
@@ -72,6 +79,10 @@ class GrpcServer {
   const std::string name_;
   /// Port of this server.
   int port_;
+  /// Flag indicates whether this server has closed.
```
```suggestion
  /// Whether this server has been closed.
```

edoakes commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -1,88 +1,62 @@
+using ResourceMappingType =
+    std::unordered_map<std::string, std::vector<std::pair<int64_t, double>>>;
 
+/// Client used for communicating with raylet.
```
```suggestion
/// Client used for communicating with the raylet.
```

edoakes commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -1,88 +1,62 @@
 class RayletClient {
  public:
-  /// Connect to the raylet.
+  /// Constructor for raylet client.
```
```suggestion
  /// Constructor for the raylet client.
```

edoakes commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -1,88 +1,62 @@
-  /// Connect to the raylet.
+  /// Constructor for raylet client.
+  /// TODO(jzh): At present, client call manager and reply handler service are generated
+  /// in raylet client. Should add parameters to the constructor and pass them in.
```
```suggestion
  /// in raylet client. Instead, we should add parameters to the constructor and pass them in.
```

Full sentences please.

edoakes commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -1,88 +1,62 @@
+  /// Send disconnect request to local raylet.
+  ray::Status Disconnect();
+
+  /// Submit task to local raylet
```
```suggestion
  /// Submit a task to the local raylet.
```

edoakes commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -1,88 +1,62 @@
-  /// \param The task specification.
-  /// \return ray::Status.
+  /// \param request The request message.
+  /// \param callback The callback function that handles reply from server
```
```suggestion
  /// \param callback The callback function that handles reply from server.
```

edoakes commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -189,7 +163,20 @@ class RayletClient {
+  /// reconnect if failed. We need this because raylet client may start before raylet
+  /// server.
+  ///
+  /// \param times Retry times.
```
```suggestion
  /// \param times Number of times to retry.
```

edoakes commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,385 @@
+                 << "unix://" + raylet_socket
+                 << ", is worker: " << (is_worker_ ? "true" : "false")
+                 << ", worker id: " << worker_id;
+  // Try to register client for 10 times.
```
```suggestion
  // Try to register client 10 times.
```

Can we make this a named constant or configurable somewhere? It's easy for magic numbers like these to get lost.

edoakes commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,385 @@
+    for (size_t i = 0; i < reply.fractional_resource_ids().size(); ++i) {
+      auto const &fractional_resource_ids = reply.fractional_resource_ids()[i];
+      auto &acquired_resources = resource_ids_[fractional_resource_ids.resource_name()];
+
```
The following block is a little bit hard to follow - a two-sentence comment here describing what it's doing would go a long way.

edoakes commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,385 @@
+}
+
+ray::Status RayletClient::RegisterClient() {
+  // Send register client request to raylet server.
```
Should this be a header comment? Can probably just remove it.

edoakes commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,385 @@
+}
+
+void RayletClient::Heartbeat() {
+  // Send register client request to raylet server.
```
Same as above.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1652/
Test FAILed.

jiangzihao2009 commented:
> Thanks, this is great!

Thank you for the comments. Remaining problem is the macro for server call factories, I will open a new PR for it.

jiangzihao2009 commented on java/runtime/src/main/java/org/ray/runtime/raylet/RayletClientImpl.java:
```diff
@@ -39,9 +39,9 @@
-  public RayletClientImpl(String schedulerSockName, UniqueId clientId,
-                          boolean isWorker, JobId jobId) {
-    client = nativeInit(schedulerSockName, clientId.getBytes(),
+  public RayletClientImpl(String schedulerSockName, UniqueId workerId,
```
I think we need this. Client id is also used in gcs client which makes things confused. All ids in raylet client or node manager for a worker are using `WorkerID`. Raylet client should be attached to a worker in the new implementation of core worker.

jiangzihao2009 commented on src/ray/raylet/worker_pool.cc:
```diff
@@ -16,21 +16,22 @@ namespace {
 
 // A helper function to get a worker from a list.
 std::shared_ptr<ray::raylet::Worker> GetWorker(
```
Yes, it's more clear to delete them.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15364/
Test FAILed.

raulchen commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,386 @@
+          *stub_, &RayletService::Stub::PrepareAsyncHeartbeat, heartbeat_request,
+          callback);
+
+  heartbeat_timer_.expires_from_now(boost::posix_time::milliseconds(300));
```
should use `heartbeat_timeout_milliseconds`

raulchen commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1299,6 +1226,67 @@ void NodeManager::ProcessSetResourceRequest(
+  bool is_worker = request.is_worker();
+  const auto worker_id = WorkerID::FromBinary(request.worker_id());
+  RAY_LOG(DEBUG) << "Received a HeartbeatRequest from worker " << worker_id
+                 << ", is_worker=" << is_worker;
```
This log is too verbose. We should remove it.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1655/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1668/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15383/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15384/
Test FAILed.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -358,8 +361,8 @@ ray::Status RayletClient::RegisterClient() {
   return GrpcStatusToRayStatus(status);
 }
 
+// Send heartbeat request to raylet server.
```
Comments that describe the method functionality should always go in the header file.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -35,15 +36,20 @@ namespace rpc {
+  /// \param[in] is_worker Indicates whether a worker or a driver.
+  /// \param[in] job_id The job id that this raylet client belongs to.
+  /// \param[in] language The language type, python or java.
+  /// \param[in] port The listening port of the worker server.
```
I know it wasn't added by you, but could you add a brief comment here explaining what the special values are (e.g., -1 means that the worker does not have a server).

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -203,6 +209,9 @@ class RayletClient {
   /// Heartbeat timer.
   boost::asio::deadline_timer heartbeat_timer_;
+
+  /// Mutex for thread safe.
```
For future reference, anytime we need to add a lock, we should think very carefully about when/whether it is necessary, and the description here should include that information.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,388 @@
+      auto const &fractional_resource_ids = reply.fractional_resource_ids()[i];
+      auto &acquired_resources = resource_ids_[fractional_resource_ids.resource_name()];
+
+      // Each resource includes a serial of ids and corresponding fractional numbers.
```
```suggestion
      // Each resource includes a series of resource IDs (e.g., GPU 0) and corresponding amount for that resource ID. If the resource amount is fractional, then there should only be one resource ID.
```

zhijunfu commented on src/ray/rpc/grpc_server.h:
```diff
@@ -31,15 +31,15 @@ class GrpcServer {
-  GrpcServer(const std::string &name, const std::string &unix_socket_path)
-      : GrpcServer(name, 0) {
+  GrpcServer(std::string name, const std::string &unix_socket_path)
+      : GrpcServer(std::move(name), 0) {
```
I previously had the same question, but actually this does the right thing and is the suggested style - it can avoid the copy if user calls the constructor with an explicit `std::move`. On the other hand, I agree that for not performance critical code like this, an extra copy is also ok, and can just use `const &`.

https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference/51706522

raulchen replied:
This change was actually introduced by my PR. Please see my comment here https://github.com/ray-project/ray/pull/5196#discussion_r303291390

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1687/
Test FAILed.

stephanie-wang commented on src/ray/raylet/worker.cc:
```diff
@@ -110,10 +107,16 @@ void Worker::AcquireTaskCpuResources(const ResourceIdSet &cpu_resources) {
+void Worker::SetGetTaskReplyAndCallback(
+    rpc::GetTaskReply *reply, const rpc::SendReplyCallback &&send_reply_callback) {
+  RAY_CHECK(reply_ == nullptr && send_reply_callback_ == nullptr);
+  reply_ = reply;
```
Hmm actually maybe I'm misunderstanding something, but why do we need to save `reply` along with `send_reply_callback`?

jiangzihao2009 replied:
Use this callback to send reply to raylet client.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15474/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15470/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15482/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15497/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15496/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1723/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1724/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15505/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15507/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1732/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15515/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1734/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15521/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1741/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15541/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15545/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15548/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1754/
Test FAILed.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,404 @@
+          *stub_, &RayletService::Stub::PrepareAsyncHeartbeat, heartbeat_request,
+          callback);
+
+  heartbeat_timer_.expires_from_now(boost::posix_time::milliseconds(300));
```
Please change this to use the config value for the timeout period.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15580/
Test FAILed.

jiangzihao2009 commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,404 @@
+
+  auto callback = [this](const Status &status, const SubmitTaskReply &reply) {
+    /// We check is_connnected_ here to avoid verbose fail logs.
+    if (!status.ok() && is_connected_) {
```
I have added a macro for it.  Check is_connected_ here because I don't want the call that has been sent to print the verbose log.

jiangzihao2009 commented on src/ray/rpc/raylet/raylet_client.h:
```diff
@@ -196,8 +195,33 @@ class RayletClient {
+  boost::asio::deadline_timer heartbeat_timer_;
+
+  /// Mutex for to protect resource_ids_.
+  mutable std::mutex resource_ids_lock_;
```
### According to the previous implementation I think the raylet client is a thread-safe client. Assuming the resource_ is using in a thread, and we are getting task in another thread and calling the clear() method, it might crash.

jiangzihao2009 replied:
Yes, I have removed it.

jiangzihao2009 commented on python/ray/tests/conftest.py:
```diff
@@ -30,7 +30,7 @@ def generate_internal_config_map(**kwargs):
     internal_config = json.dumps({
         "initial_reconstruction_timeout_milliseconds": 200,
-        "num_heartbeats_timeout": 10,
+        "num_heartbeats_timeout": 50,
```
ok, these is a problem caused by post() function introduced in grpc server. Too many post events block the heartbeat timer and the raylet is marked as dead when there are lots of task. I am trying to solve this problem when this config stays the same.

jiangzihao2009 commented on src/ray/protobuf/raylet.proto:
```diff
@@ -0,0 +1,205 @@
+  // timeout
+  int64 timeout = 3;
+  // Whether to wait until objects appear locally.
+  bool wait_local = 4;
```
wait_local means wait for a local objects, all variables related to this in raylet is wait_local. And the variable in prevous raylet client is also wait_local, I think we should keep it unchanged.

> "wait_local" is somewhat unclear to me - sounds like it might specify if the client or server does the waiting or something like that.

wait_local means wait for a local objects, all variables related to this in raylet is wait_local. And the variable in prevous raylet client is also wait_local, I think we should keep it unchanged.

raulchen commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -0,0 +1,400 @@
+  // Callback to deal with reply.
+  auto callback = [this](const Status &status, const FetchOrReconstructReply &reply) {
+    if (!status.ok() && is_connected_) {
+      is_connected_ = false;
```
For other async requests, we should also check `call->GetStatus()` and set `is_connected_` to false if not ok.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1757/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15582/
Test FAILed.

jiangzihao2009 commented:
> Just some minor comments.
> 
> Can you ping me when this is ready for a final review? Let's merge this as soon as the last comments are addressed and the tests pass.

No problem@stephanie-wang, I have addressed these comments and tests.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15585/
Test PASSed.

stephanie-wang commented on src/ray/rpc/raylet/raylet_client.cc:
```diff
@@ -131,11 +126,13 @@ ray::Status RayletClient::GetTask(std::unique_ptr<ray::TaskSpecification> *task_
 
-  auto callback = [](const Status &status, const TaskDoneReply &reply) {
-    if (!status.ok()) {
+  auto callback = [this](const Status &status, const TaskDoneReply &reply) {
```
One option to help make sure we're consistent with the reply callbacks is to add a templated helper function to return the callback, something like this:
```c++
template <typename ReplyT>
std::function<void(const &Status, const &ReplyT)> MakeReplyCallback() {
  auto callback = ...
  return callback;
}
```

jiangzihao2009 replied:
It seems not necessary, because we still need to add the actual content of callback.

stephanie-wang commented:
Let me know if you need help debugging the Travis failures (they look like they're due to this PR?).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15605/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15606/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15613/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15607/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15614/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15615/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15618/
Test PASSed.

jiangzihao2009 commented:
> Let me know if you need help debugging the Travis failures (they look like they're due to this PR?).

At present, it seems there are two failures. 
First is the fork_consistency in test_actor.py caused by not enough memory problem(msg like this ValueError: The default object store size of 1.29 GB will use more than 90% of the available memory on this node (0.78 GB)). It seems changing to GRPC cost more memory than asio library. It might just because the release of memory is not fast and it will pass if I only run this single task.
Secondly it crashes in function UnsubscribeGetDependencies when running test  test_multi_node.py::test_driver_exiting_when_worker_blocked. It might be caused by double removing the dependencies due to driver exitting.
The first one seems not introduced by this PR, maybe we should change the algorithm for calculating available memory. The second one seems related to the assign task introduced in this pr.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15619/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1777/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15635/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/15636/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1778/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1779/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1787/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1788/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1789/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1792/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1793/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1808/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/1810/
Test FAILed.
